core/src/main/clojure/xtdb/expression.clj

```diff
@@ -1375,6 +1375,33 @@
   {:return-type :i32
    :->call-code #(do `(.size ~@%))})
 
+(defmethod codegen-call [:length :utf8] [expr]
+  (codegen-call (assoc expr :f :character_length)))
+
+(defmethod codegen-call [:length :varbinary] [expr]
+  (codegen-call (assoc expr :f :octet_length)))
+
+(defmethod codegen-call [:length :list] [expr]
+  (codegen-call (assoc expr :f :cardinality)))
+
+(defn size-of-set [^Set set] 
+  (.size ^Set set))
+
+(defmethod codegen-call [:length :set] [_] 
+  {:return-type :i32
+   :->call-code #(do `(size-of-set ~@%))})
+
+(defn count-non-empty [m]
+  (reduce
+   (fn [^Integer non-empty [_x ^IValueReader y]]
+     (if (.isNull y) non-empty (Math/addExact non-empty 1)))
```

Comment:
Originally done to cover "non empty field" requirement - nils are a value and as such are important to count - rather, we want to check here for **absent** (ie, there is no field).

Reply:
Will need to implement `isAbsent` on IValueReader and it's implementers.